cli/cliui/table.go

```diff
@@ -41,3 +44,174 @@ func FilterTableColumns(header table.Row, columns []string) []table.ColumnConfig
 	}
 	return columnConfigs
 }
+
+// DisplayTable renders a table as a string. The input argument must be a slice
+// of structs (panics otherwise, even for empty slices). At least one field in
+// the struct must have a `table:""` tag containing the name of the column in
+// the outputted table.
+//
+// Nested structs are processed if the field has the `table:"$NAME,recursive"`
+// tag and their fields will be named as `$PARENT_NAME $NAME`. If the tag is
+// malformed or a field is marked as recursive but does not contain a struct or
+// a pointer to a struct, this function will panic (even with an empty input
+// slice).
+//
+// If sort is empty, the input order will be used. If filterColumns is empty or
+// nil, all available columns are included.
+func DisplayTable(out interface{}, sort string, filterColumns []string) (string, error) {
```

Comment:
To me most of the `panic`s look like input validation and could be bubbled up, so I think it would be nice to return errors here instead (also considering the error return signature is already present). Would give more control to the caller (and if a user ever hit this, a nice error message vs "explode in face" 😄). Or alternatively `recover` here, although that could mask the trace to some actual unhandled cases.

Also: https://github.com/golang/go/wiki/CodeReviewComments#dont-panic

Reply:
These panics are only for developer errors which are absolutely bugs (I.e. they didn't use the table struct tags properly). Since the validation happens at the type level, it even happens when there are no items in the input slice.

Reply:
I don't see developer errors/bugs automatically translating into `panic`s. IMO there should be a very high threshold for introducing `panic`s (especially if they happen at runtime) and there should be some benefit over regular error handling. Even though this behavior is documented quite well, a reader of the code using this function would not know that there is a hidden contract where this function might not exit cleanly (with an error), so they couldn't infer that the `if err != nil` case won't run, for instance.

So for the sake of argument, let's assume that at some point in the future, we expand this function with some functionality and as a result make it a bit more strict. We also have a code-path that is rarely hit (and unfortunately, not tested). Let's say, sometimes we display user info with additional data (wrap them in extended structs). Then, after we've made this change, only the extended user info fails the validations in `DisplayTable`, which is only hit by a subset of our users. It wouldn't be the end of the world but it'd be a nicer user experience to see an error message vs stack trace.

A contrived example, but should illustrate why I feel `panic`s at runtime are a bad idea.

Reply:
> Let's say, sometimes we display user info with additional data (wrap them in extended structs). Then, after we've made this change, only the extended user info fails the validations in DisplayTable, which is only hit by a subset of our users

This cannot happen because the types themselves are checked, even if nil or if there are empty slices. For example:

```go
type x struct {
  y *y `table:"y,recursive"`
}

type y struct {
  // bad table tag should cause a panic
  name string `table:""`
}

func main() {
  // full data
  _, _ = DisplayTable([]x{{y: &y{name: "hi"}}}, ...) //panics

  // nil inner field still checks inner type
  _, _ = DisplayTable([]x{{y: nil}}, ...) // panics

  // empty slice still checks inner type
  _, _ = DisplayTable([]x{}, ...) // panics
}
```

I think that panics are suitable for developer errors, and that it would be very difficult for a panic like this to hit production because of what I described above unless there were no tests that hit that line of code (which seems unlikely).

However I'm fine with just returning an error instead since there's already an error return type. I'll change the code to error instead of panic

Reply:
All panics are now errors so I'm resolving this thread. The functionality for always checking types recursively even if empty or nil remains though which mitigates most of the risk.